### PR TITLE
[IMP] mail: improve field tracking

### PR DESCRIPTION
**Pull request by rco-odoo**
Improve tracking by generating the tracking message at flush.
Also implement the tracking for stored computed fields.

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -539,20 +577,14 @@ def message_change_thread(self, new_thread):
...
-                tracked_fields.append(name)
-
-        if tracked_fields:
-            return self.fields_get(tracked_fields)
-        return {}
+        """ Return the set of tracked fields names for the current model. """
+        return {
+            name
+            for name, field in self._fields.items()
+            if getattr(field, 'tracking', None) or getattr(field, 'track_visibility', None)
```
Comment (seb-odoo) [inline addons/mail/models/mail_thread.py]: Shouldn't we finally deprecate/remove `track_visibility`?

File: odoo/models.py [hunk 2]
```diff
@@ -5397,10 +5397,19 @@ def process(model, id_vals):
 
         if fnames is None:
             # flush everything
-            self.recompute()
-            while self.env.all.towrite:
-                model_name, id_vals = self.env.all.towrite.popitem()
-                process(self.env[model_name], id_vals)
+            for i in range(10):
```
Comment (seb-odoo) [inline odoo/models.py]: This seems so arbitrary.

File: addons/mail/models/mail_thread.py (hunk 1 above)
Comment (rco-odoo) [reply addons/mail/models/mail_thread.py]: Yes, you can do that :wink:

File: addons/mail/models/mail_thread.py [hunk 3]
```diff
@@ -539,20 +577,14 @@ def message_change_thread(self, new_thread):
     # Automatic log / Tracking
     # ------------------------------------------------------
 
-    @api.model
+    @tools.ormcache()
```
Comment (dbo-odoo) [inline addons/mail/models/mail_thread.py]: are you sure everything gets invalidated properly when modifying custom fields; e.g. writing on `tracking` in `ir.model.fields`?

Comment (rco-odoo) [reply addons/mail/models/mail_thread.py]: Yes, there is a call to `clear_caches` when modifying a custom field.

**Comment by dbo-odoo** [comment]
@rco-odoo 
is it possible to get this for 13.2? You discussed it last week with @arj-odoo, he based his branch on this for subscription changes tracking... :)

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**